storage/remote/sigv4.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/textproto"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	signer "github.com/aws/aws-sdk-go/aws/signer/v4"
+	"github.com/prometheus/prometheus/config"
+)
+
+var sigv4HeaderDenylist = []string{
+	"uber-trace-id",
+}
+
+type sigV4RoundTripper struct {
+	region string
+	next   http.RoundTripper
+	pool   sync.Pool
+
+	signer *signer.Signer
+}
+
+// newSigV4RoundTripper returns a new http.RoundTripper that will sign requests
+// using Amazon's Signature Verification V4 signing procedure. The request will
+// then be handed off to the next RoundTripper provided by next. If next is nil,
+// http.DefaultTransport will be used.
+//
+// Credentials for signing are retrieved using the the default AWS credential
+// chain. If credentials cannot be found, an error will be returned.
+func newSigV4RoundTripper(cfg *config.SigV4Config, next http.RoundTripper) (http.RoundTripper, error) {
+	if next == nil {
+		next = http.DefaultTransport
+	}
+
+	creds := credentials.NewStaticCredentials(cfg.AccessKey, string(cfg.SecretKey), "")
+	if cfg.AccessKey == "" && cfg.SecretKey == "" {
+		creds = nil
+	}
+
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config: aws.Config{
+			Region:      aws.String(cfg.Region),
```

Comment:
This doesn't seem to be honoring the default AWS region.

If I run Prometheus without any region set in the config, I'm getting 403. The following configuration is the only way I was able to successfully send data:

```
remote_write:
  - url: https://aps-workspaces.us-west-2.amazonaws.com/workspaces/ws-xxx-xxx-xxx-xxx-0b09141a9e9e/api/v1/remote_write
    sigv4:
      region: us-west-2
````  

My default AWS region is already "us-west-2":

```
$ aws configure get region
us-west-2
```

Reply:
Just checking: did you try with `sigv4: {}` ?
`sigv4:` alone is not enough.

Reply:
Ah, thanks for pointing this out. It looks like it's not using the shared configuration file by default and is expecting `AWS_SDK_LOAD_CONFIG=1` to be set. The EC2 SD configs seem to have the same problem. 

What would you recommend the default here to be?

Reply:
It seems that passing the region as env var also works (AWS_REGION). I think we could look for the env var and if not found use  https://github.com/aws-observability/amp-eks-iam/blob/main/main.go#L141 ?

Reply:
Just FYI, https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials documents on the credential resolution process for AWS Go SDK v1. 

You might know this already, but AWS has a v2 SDK (which is supposed to be better): https://github.com/aws/aws-sdk-go-v2. The credential resolution process for v2 sdk can be found here: https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/#specifying-credentials

The marketing text for AWS SDK v2 for Go can be found at: https://aws.amazon.com/blogs/developer/aws-sdk-for-go-version-2-general-availability/

Reply:
@rfratto , did you have a chance to look at this?

Reply:
My gut feeling is we should defer exploring v2 of the AWS SDK to a future PR that updates both this and the EC2 SD at the same time. 

Mostly I want to avoid having this PR linger on, but I'm also nervous about subtly breaking things by switching to a new major version of the AWS lib.

Reply:
https://github.com/aws-observability/amp-eks-iam/blob/main/main.go#L141  This is using the SDK v1 if I see correctly?

Reply:
But I agree that we could get this in ; it matches the EC2 SD config ; we can open an issue for the default SDK region and figure it out.